src/components/route-book/index.tsx

```diff
@@ -0,0 +1,175 @@
+  const totals = new Map<number, RouteWithTotal>();
+
+  data.forEach((route: Route) => {
+    const entry = totals.get(route.price);
+    if (entry) {
+      totals.set(route.price, {
+        ...route,
+        total: entry.total + route.amount,
+        amount: Math.min(entry.amount, route.amount),
+      });
+    } else {
+      totals.set(route.price, {
+        ...route,
+        total: route.amount,
+      });
+    }
+  });
+
+  return isBuySide
+    ? Array.from(totals.values())
+        .slice(0, limit)
+        .toSorted((a, b) => b.price - a.price)
+    : Array.from(totals.values()).slice(-limit);
+}
+
+function getDisplayData({
+  data,
+  computePositionId,
+  asset1,
+  asset2,
+  isBuySide,
+  limit,
+}: {
+  data: Position[];
+  computePositionId: ((position: Position) => PositionId) | undefined;
+  asset1: Metadata | undefined;
+  asset2: Metadata | undefined;
+  isBuySide: boolean;
+  limit: number;
+}): RouteWithTotal[] {
+  if (!computePositionId || !asset1 || !asset2) {
+    return [];
+  }
+
+  const asset1Exponent = getDisplayDenomExponent(asset1);
+  const asset2Exponent = getDisplayDenomExponent(asset2);
+
+  const getValue = (property: Amount | undefined, exponent: number) =>
+    fromBaseUnit(BigInt(property?.lo ?? 0), BigInt(property?.hi ?? 0), exponent);
+
+  const routes = data
+    // .filter(position => position.state?.state.toLocaleString() === 'POSITION_STATE_ENUM_OPENED')
+    .filter(position => position.state?.state === PositionState_PositionStateEnum.OPENED)
+    .map(position => {
+      const direction = position.phi?.pair?.asset2?.equals(asset1.penumbraAssetId) ? -1 : 1;
+
+      const r1 = getValue(position.reserves?.r1, direction === 1 ? asset1Exponent : asset2Exponent);
+      const r2 = getValue(position.reserves?.r2, direction === 1 ? asset2Exponent : asset1Exponent);
+      const p = getValue(
+        position.phi?.component?.p,
+        direction === 1 ? asset2Exponent : asset1Exponent,
+      );
+      const q = getValue(
+        position.phi?.component?.q,
+        direction === 1 ? asset1Exponent : asset2Exponent,
+      );
+      const price = Number(direction === 1 ? p.div(q) : q.div(p));
+      const amount = isBuySide
+        ? Number(direction === 1 ? r2 : r1) / price
+        : Number(direction === 1 ? r1 : r2);
+
+      const id = computePositionId(position);
+      const innerStr = uint8ArrayToBase64(id.inner);
+      const bech32Id = innerToBech32Address(innerStr, 'plpid');
+
+      return {
+        lpId: bech32Id,
+        position,
+        price,
+        amount,
+      };
+    })
+    .filter(displayData => displayData.amount > 0)
+    .toSorted((a, b) => b.price - a.price) as Route[];
+
+  return getTotals(routes, isBuySide, limit);
+}
+
+export function RouteBook() {
+  const { data: assets } = useAssets();
+  const asset1 = assets?.find(asset => asset.symbol === 'UM');
+  const asset2 = assets?.find(asset => asset.symbol === 'GM');
```

Comment:
suggestion: use `pair` MobX state (`src/shared/state/pair.ts`) to get the asset info. You may also fill this state initially, so `from` and `to` won't be `undefined`

Reply:
This doesn't seem to work for me, the asset selector doesn't seem to fire on click?

Reply:
That's weird, it works fine for me... Would love if you can make the reproduction

https://github.com/user-attachments/assets/49067ddb-1ed5-4fce-8542-d2b6fb273e6b